### DISCUSSION TITLE
ESP32 Animated GIF Fails when loaded from SD card

### DISCUSSION THREAD

**Discussion by timmgleason** [moononournation/Arduino_GFX#104] [Q&A] [answered]
Moving on from the Seeeduino XIAO, I've gotten "ImgViewerAnimatedGIF" working on an ESP32 dev board w/ 4MB PSRAM.
When I load the images into the onboard PSRAM and run the sketch, it loads the GIF and plays in an endless loop.
Exactly as it is supposed to work.

Now when I put the images onto an SD card, it will load the GIF, display one frame and error out. To attempt again, one must remove the SD card, reset the ESP32 board and re-insert the SD card. It never plays more than a single frame, usually with 'tears' in the image display (scrambled artifacts that are not in the actual GIF). It matters not whether using the included GIFs or cropped and optimized other GIFS.

Here are the changes I've made to the example sketch (from commit  57a7cb1):

Line 56-70;
> #define TFT_CS  5 // -1 for display without CS pin
> #define TFT_DC 27
> #define TFT_RST 33
> #define TFT_BL 22
> #define TFT_MOSI 23
> #define TFT_CLK 18
> Arduino_DataBus *bus = new Arduino_ESP32SPI(TFT_DC /* DC */, TFT_CS /* CS */, TFT_CLK /* SCK */, TFT_MOSI /* MOSI */, -1 /* MISO */, VSPI /* spi_num */);
> Arduino_GFX *gfx = new Arduino_GC9A01(bus, TFT_RST, 0 /* rotation */, true /* IPS */);
> 
> /* More dev device declaration: https://github.com/moononournation/Arduino_GFX/wiki/Dev-Device-Declaration */
> #if defined(DISPLAY_DEV_KIT)
> Arduino_GFX *gfx = create_default_Arduino_GFX();
> #else /* !defined(DISPLAY_DEV_KIT) */
> 
> /* More data bus class: https://github.com/moononournation/Arduino_GFX/wiki/Data-Bus-Class */
> // Arduino_DataBus *bus = create_default_Arduino_DataBus();
> 
> /* More display class: https://github.com/moononournation/Arduino_GFX/wiki/Display-Class */
> // Arduino_GFX *gfx = new Arduino_ILI9341(bus, TFT_RST, 3 /* rotation */, false /* IPS */);
> 
> #endif /* !defined(DISPLAY_DEV_KIT) */

Lines 123-127;
> #elif defined(ESP32)
>   // if (!FFat.begin())
>   // if (!LittleFS.begin())
>   //if (!SPIFFS.begin())
>   if (!SD.begin(TFT_CS))

Lines 146-150;
> #elif defined(ESP32)
>     // File gifFile = FFat.open(GIF_FILENAME, "r");
>     // File gifFile = LittleFS.open(GIF_FILENAME, "r");
>     //File gifFile = SPIFFS.open(GIF_FILENAME, "r");
>     File gifFile = SD.open(GIF_FILENAME, "r");

Loading from the onboard SPIFFS memory produces on the console "GIF video start" and the animation runs.
Loading from the SD card produces the following on the console:

> new_table() malloc: 32776
> GIF video start
> Read sep: [⸮].
> 
> ERROR: gd_get_frame() failed!
> GIF video end
> duration: 100, remain: 4294967146 (ovf%)

I can run any number of other SD card sketches that read and write via the ESP32 board I have, it's just the animated GIFs that seems to have an issue when playing. 

NOTE * A nearly identical thing happens with the other Img exmaples as well when reading from the SD card.
For example, the "ImgViewerPng" will work properly from the PSRAM using SPIFFS, but will load a partial image only one time when loading from the SD card. So instead of the initial load, followed by repeated offset and reduced loadings, it loads part of the image and then just stops.

I am using a new SanDisk  Ultra 32GZB Class 10 SDHC card, if that makes any difference.

Thanks!

**Comment by moononournation** [comment, marked as answer]
If you want share SD and display with same SPI, you need to use individual CS pin for each device. You may find more detail in this example code:
https://github.com/moononournation/RGB565_video/blob/master/SD_GIF_video/SD_GIF_video.ino

**Reply by timmgleason** [reply]
I suspected something like that, but I am unfortunately new to this embedded Arduino programming and haven't found a good primer yet. So it's trial and error learning.
I'll give that a look over and I appreciate the assistance and pointers.
Cheers!

**Reply by timmgleason** [reply]
Bingo!

Defined an SS value and set that inside the SD.begin call, works like a charm.

So to summarize;
Lines 65-62:
> #define TFT_CS  5 // -1 for display without CS pin
> #define TFT_DC 27
> #define TFT_RST 33
> #define TFT_BL 22
> #define TFT_MOSI 23
> #define TFT_CLK 18
> **#define SS 4**

Lines 123-127:

> #elif defined(ESP32)
>   // if (!FFat.begin())
>   // if (!LittleFS.begin())
>   //if (!SPIFFS.begin())
>   if (!SD.begin(SS))

Cheers
